stream-chat-android-client/src/main/java/io/getstream/chat/android/client/network/NetworkStateProvider.kt

```diff
@@ -0,0 +1,50 @@
+package io.getstream.chat.android.client.network
+
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkRequest
+
+internal class NetworkStateProvider(private val connectivityManager: ConnectivityManager) {
+
+    private val callback = object : ConnectivityManager.NetworkCallback() {
+        override fun onAvailable(network: Network?) {
+            if (!isConnected) {
+                isConnected = true
+                listener?.onConnected()
+            }
+        }
+
+        override fun onLost(network: Network?) {
+            // Checks whether the network was switched or connection was lost
+            val isConnected = connectivityManager.activeNetworkInfo?.isConnected ?: false
+            if (!isConnected && this@NetworkStateProvider.isConnected) {
+                this@NetworkStateProvider.isConnected = false
+                listener?.onDisconnected()
+            }
+        }
+    }
+
+    private var isConnected: Boolean = false
```

Comment:
I think it's better to use `@Volatile` or `AtomicBoolean` here instead of simple property.

Reply:
Marking with @Volatile

Reply:
Marking with Volatile